staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -197,7 +197,7 @@ func (tw *baseTimeoutWriter) WriteHeader(code int) {
 	tw.mu.Lock()
 	defer tw.mu.Unlock()
 
-	if tw.timedOut || tw.wroteHeader || tw.hijacked {
+	if tw.timedOut || (tw.wroteHeader && code != http.StatusInternalServerError) || tw.hijacked {
```

Comment:
I think a comment explaining the purpose of this would be helpful.

A test would be nice, too :)

Reply:
+1

Reply:
What about when the connection has been closed by the other end? 
Also please take a look at https://github.com/kubernetes/kubernetes/pull/75887/files.
I believe that fixes explicitly hijacks connections on proxy to get around a problem flushing headers. 
This will be no worse in that case, just not sure it will help.

Reply:
> What about when the connection has been closed by the other end?

This error should be logged properly with this change.  This was the error case that I based my testing off of in #75983.  My understanding is that if the connection has been closed on the other end, a `Write` will return a 'broken pipe' which is caught by ErrorToAPI.  The issue before was that the timeout handler appears to have tw.WroteHeader set in this case but it seems the header written was a 200 while a 500 is passed into 'WriteHeader'.

> Also please take a look at https://github.com/kubernetes/kubernetes/pull/75887/files.
I believe that fixes explicitly hijacks connections on proxy to get around a problem flushing headers.
This will be no worse in that case, just not sure it will help.

Nice! This definitely is a very similar case.  Are you suggesting perhaps hijacking should be used in the timeout handler?  The issue I was witnessed here was that tw.WroteHeader is being set either in another handler (I'm not sure how this is possible) or in a series of consecutive Write calls.  This if case seemed to solve the direct issue of not logging things from ErrorToAPI.

Reply:
> A test would be nice, too :)

I'm on it!  I had some trouble writing a test that reproduces the entirety of the original issue but I think adding a test in timeout_test.go that gets a 500 and verifies that a 500 shows in the response makes sense here.

Reply:
Not sure when we log the status code on a proxy request. But I'm assuming its after the flushHeaderWriter has hijacked the connection. My concern then was that the hijack flag was going to be true and you will still drop into the if statement and hence return.

Reply:
In further investigation, it makes sense to also do this in the case of hijacking.  The issue here is that in all cases where is a Write/WriteHeader (Write is most likely) command is called and then a 500 occurs, the timeout handler masks this 500 as it assumes the header being written has been taken care of from the first write although it has changed.  This has the issues of hiding the 500 from the httplogger and logs.